src/schema.rs

```diff
@@ -0,0 +1,462 @@
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use snafu::{ensure, OptionExt};
+
+use crate::error::{NoTypesSnafu, Result, UnexpectedSnafu};
+use crate::proto;
+
+use arrow::datatypes::{DataType as ArrowDataType, Field, Schema, TimeUnit, UnionMode};
+
+/// Represents the root data type of the ORC file. Contains multiple named child types
+/// which map to the columns available. Allows projecting only specific columns from
+/// the base schema.
+///
+/// This is essentially a Struct type, but with special handling such as for projection
+/// and transforming into an Arrow schema.
+///
+/// Note that the ORC spec states the root type does not necessarily have to be a Struct.
+/// Currently we only support having a Struct as the root data type.
+///
+/// See: <https://orc.apache.org/docs/types.html>
+#[derive(Debug, Clone)]
+pub struct RootDataType {
+    children: Vec<(String, DataType)>,
+}
```

Comment:
Maybe in future we can add handling for files with no struct as the root (if these exist?)

Probably just involves a bit of extra checks and field name generation, e.g. col1, col2, ...

Reply:
Can we replace `RootDataType ` with `DataType ` to support no struct as the root?

Reply:
Perhaps.

I just found it a bit clunky since had to account for if Struct (has names) or if is not struct then would need to create the names, etc.

Could just leave as future work, as again I'm unsure if such files are commonly found (no struct as root type)